mmengine/hooks/checkpoint_hook.py

```diff
@@ -304,6 +325,58 @@ def after_val_epoch(self, runner, metrics):
 
         self._save_best_checkpoint(runner, metrics)
 
+    def after_train(self, runner) -> None:
+        """Publish the checkpoint after training.
+
+        Args:
+            runner (Runner): The runner of the training process.
+        """
+        if self.published_keys is None:
+            return
+
+        if self.save_last and 'last_ckpt' in runner.message_hub.runtime_info:
+            last_ckpt = runner.message_hub.get_info('last_ckpt')
+            self._publish_model(runner, last_ckpt)
+
+        if getattr(self, 'best_ckpt_path', None) is not None:
+            self._publish_model(runner, str(self.best_ckpt_path))
+        if getattr(self, 'best_ckpt_path_dict', None) is not None:
+            for key, best_ckpt in self.best_ckpt_path_dict.items():
+                self._publish_model(runner, best_ckpt)
+
+    def _publish_model(self, runner, ckpt_path: str) -> None:
+        """Remove unnecessary keys from ckpt_path and save the new checkpoint.
+
+        Args:
+            runner (Runner): The runner of the training process.
+            ckpt_path (str): The checkpoint path that ought to be published.
+        """
+        from mmengine.runner import save_checkpoint
+        from mmengine.runner.checkpoint import _load_checkpoint
+        checkpoint = _load_checkpoint(ckpt_path)
+        assert self.published_keys is not None
+        removed_keys = []
+        for key in list(checkpoint.keys()):
+            if key not in self.published_keys:
+                removed_keys.append(key)
+                checkpoint.pop(key)
+        if removed_keys:
+            print_log(
+                f'Key {removed_keys} will be removed because they are not '
+                'found in published_keys. If you want to keep them, '
+                f'please set `{removed_keys}` in published_keys',
+                logger='current')
+        final_path = osp.splitext(
+            ckpt_path)[0] + f'-published-{runner.timestamp}.pth'
+        save_checkpoint(checkpoint, final_path)
+        print_log(
+            f'The published model is saved at {final_path}.', logger='current')
+        if 'publish_ckpt_names' not in runner.message_hub.runtime_info:
+            runner.message_hub.update_info('publish_ckpt_names', [final_path])
+        else:
+            ckpt_path_list = runner.message_hub.get_info('publish_ckpt_names')
+            ckpt_path_list.append(final_path)
```

Comment:
Is it necessary to add those messages to message_hub?

Reply:
For check in test_checkpoint_hook.py, because I could load self.best_ckpt_path in it

Reply:
We should not modify the code logic just for writing unit tests. A more reasonable approach is to check the checkpoint of the published model has been generated.

Reply:
> We should not modify the code logic just for writing unit tests. A more reasonable approach is to check the checkpoint of the published model has been saved.

oki, already rm the unnecessary messages add in message_hub.